V2 content (#357)
enviodev/docs: docs/HyperIndex/migration-guide-v1-v2.md

---
id: migration-guide-v1-v2
title: Migration Guide v1 to v2
sidebar_label: Migration Guide v1 to v2
slug: /migration-guide-v1-v2
---

NOTE: v2 is currently in rc phase (release candidate) and is not yet stable. Please refer to the [v1 documentation](https://docs.envio.dev/docs/HyperIndex) for the stable version.

# Introduction

V2 of HyperIndex is about streamlining the process of starting an indexer and optimizing it as you go. There are two big changes:

> - Handlers are now asynchronous, and by default 'loaders' aren't required at all - there is no need for the [async-mode](/docs/HyperIndex/async-mode).

Reviewer: ```suggestion
- Handlers are now asynchronous, and `loaders` became an optional tool for additional optimizations.
- It made [async-mode](/docs/HyperIndex/async-mode) not needed, hence it's removed in v2.
```

- Loaders (when used) are more expressive and connected via the return type to the context of the handler.
  - In v1, you needed to use linked entities to load entity fields of other entities. This was unintuitive.
    - In v2, you can directly access the fields of the loader the exact same way as you do in the handler, with an async 'get' function.
  - In v1, you needed to call 'load' in the loader, and 'get' in the handler separately (or use labelled fields).
    - In v2, you can use the return type of the loader to directly access the fields in the handler via the context, or you can call 'get' again.

## Changes to Make

### Handlers

> - Handlers are now asynchronous (you can add an `async` keyword before it).

Reviewer: ```suggestion
- Handlers are now asynchronous - add the `async` keyword and rename `handlerAsync` to `handler`.
```

- You can use `handlerWithLoader` if you need a loader, otherwise use `handler` directly.
- The 'get' function is now asyncronous, so add an `await` before those functions.
- No labelled entities.

### Loaders

- Loaders are merged into the handlers using `handlerWithLoader`.

> - Loading linked entities is done directly with promices in the loader.

Reviewer: ???
```suggestion
- Loading linked entities is done directly with promises in the loader.
```

- Loaders are completely optional - only use the if you care about high throughput indexing.
- Loaders return the required entities which are then used in the handler.

> - `contractRegister` is used for dynamic contract registration. You don't add contracts to the indexer dynamically in the loader anymore.

Reviewer: ```suggestion
- The dynamic contract registration moved from loaders to its own `<ContractName>.<EventName>.contractRegister` handler.
```

- The return type of the loader is used directly in the handler to access the loaded data. No need to re-'get' it again in the handler.

### Configuration

- There is no async-mode anymore, so you can remove `isAsync: true` from each of the events in your `config.yaml`.

```diff
- isAsync: true
```

> - No entity labels or required entities.

Reviewer: ```suggestion
- Removed entity labels and required entities.
```

### Miscellaneous breaking changes and deprecations

> - For rescript, you need to use the built in bigint type instead of the `Ethers.BigInt` type.

Reviewer: ```suggestion
- For ReScript, we moved to the built-in `bigint` type instead of the `Ethers.BigInt.t`.
```

- Upgrade to rescript 11 uncurried mode (and use of newer rescript compiler).
- The `context.Entity.load` function is deprecated and should be replaced with direct calls to `context.Entity.get` in the loader.
- The `context.ParentEntity.loadField` functions are deprecated and should be replaced with direct calls to `context.ChildEntity.get`.

> <!-- TODO: lots more to put here -->

Reviewer: Removed `Contract` and `Entity` suffixes among the ones I remember

Author: ```suggestion
- Remove the `Contract` and `Entity` suffixes from generated code.
<!-- TODO: lots more to put here -->
```

## Migration Steps

### 1. Update Imports

Replace the old import statements with the new ones.

**Before:**

```typescript
import {
  SomeContract.Event1.loader,
  SomeContract.Event1.handler,
  // or you aren't using these `_` versions of the imports
  SomeContract,
  // ...
} from "../generated/src/Handlers.gen"; // Not all imports still look like this, but on old indexers they do.

import {
  Event1Entity,
  Event2Entity,
  // ... other entities
} from "../generated/src/Types.gen";
```

**After:**

```typescript
import {
  Some,

>   Another,

Author: Hmm, good feedback. Maybe make it say 'Greeter' and 'User' or something like that?

SomeContract.Event1.loader(({ event, context }) => {
  // Loader code
});
SomeContract.Event1.handler(({ event, context }) => {
  // Handler code
});
```

**After:**

```typescript
SomeContract.Event1.handlerWithLoader({
  loader: async ({ event, context }) => {
    // Loader code
    return {
      /* loaded data */
    };
  },
  handler: async ({ event, context, loaderReturn }) => {
    // Handler code using loaderReturn
  },
});
```

#### Or without a loader:

**Before:**

```typescript
SomeContract.Event1.handler(({ event, context }) => {
  // Handler code
});
```

**After:**

```typescript
SomeContract.Event1.handler(async ({ event, context, loaderReturn }) => {
  // Handler code
});
```

### 3. Dynamic Contract Registration

Use `contractRegister` for dynamic contract registration.

**Before:**

```typescript
SomeContract.Event1.loader(({ event, context }) => {
  context.contractRegistration.addContract(event.params.contract);
});
```

**After:**

```typescript
SomeContract.Event1.contractRegister(({ event, context }) => {
  context.addContract(event.params.contract);
});
```

### 4. Handling Entities

**Before == After**

```typescript

> const entityInstance: EntityType = {

Author: You're right. Will update
